get_checks.py

```diff
@@ -0,0 +1,189 @@
+    context_lines.insert(abs(context_range[0]), "# XXXXXXXXXXXXX MARKED WARNING XXXXXXXXXXXXX")
+    context_lines.insert(abs(context_range[0]) + 2, "# XXXXXXXXXXXXX MARKED WARNING XXXXXXXXXXXXX")
+    context = literal_str("\n".join(context_lines))
+    return {'filename': fname,
+            'line_number': idx + 1,
+            'text': literal_str("\n".join(warn_text)),
+            'context': context}
+
+
+def get_sc():
+    """
+    For a list of source files, get all the indexes of lines that match
+    standard warning patterns, and fetch the context about and warning that starts
+    with the matching line.
+    """
+    perl_files = ['src/starcheck.pl',
+                  'src/lib/Ska/Starcheck/Obsid.pm',
+                  'src/lib/Ska/Starcheck/FigureOfMerit.pm',
+                  'src/lib/Ska/Parse_CM_File.pm']
+    checks = []
+    for f in perl_files:
+        text = open(f).readlines()
+        for idx, line in enumerate(text):
+            if re.match("^\s*#", line):
+                continue
+            if (re.search('[^\$]warning\s*\(', line)
+                or re.search("\$warn\s*=", line)
+                or re.search("push @\S*warn", line)
+                or re.search("push @\{\$error\}", line)
+                or re.search("push @\S*fyi", line)
+                ):
+                    checks.append(get_single_warn(f, text, idx))
+    return checks
+
+
+def add_github_urls(checks, sha):
+    """
+    Given the array of warnings/checks, determine the github URL and add to the
+    per-warning dict.
+    """
+    url = "https://github.com/sot/starcheck/blob/"
+    for check in checks:
+        check['github_url'] = "{url}{sha}/{filename}#L{line_number}".format(
+            url=url, filename=check['filename'], sha=sha, line_number=check['line_number'])
+
+
+def copy_manual_fields(c, pc):
+    """
+    Copy the manually-entered fields from a previous file into new output.
+    """
+    for field in SAVE_FIELDS:
+        if field in pc.keys():
+            c[field] = pc[field]
+            pc['copied'] = 1
+
+
+def assign_ids(checks):
+    prev_ids = [int(c['id']) for c in checks if 'id' in c]
+    prev_max = 0
+    if len(prev_ids):
+        prev_max = 1 + max(prev_ids)
+    no_id_checks = [c for c in checks if 'id' not in c]
+    for idx, c in enumerate(no_id_checks):
+        c['id'] = prev_max + idx
+        print("Assigning new id {} to check {}".format(c['id'], c['text']))
+
+
+def preserve_manual_content(checks, previous_checks):
+    """
+    For each new check, see if it was in a previous version of the output
+    and call the routine to copy over manual content.
+    """
+    for c in checks:
+        for pc in previous_checks:
+            if (pc['text'] == c['text']):
+                pc['has_match'] = 1
+                if 'id' not in c:
+                    c['id'] = pc['id']
+                    copy_manual_fields(c, pc)
+                else:
+                    print("Conflict on id {}".format(c['id']))
```

Comment:
FYI, I was amused to see that we actually do have a "conflict" here as we presently have one warning that is called twice from the code with the same text (in two different contexts).